controllers/onload_controller.go

```diff
@@ -78,34 +90,515 @@ func (r *OnloadReconciler) Reconcile(
 	}
 
 	return ctrl.Result{}, nil
 }
 
+const kmmModuleLabelPrefix = "kmm.node.kubernetes.io/version-module"
+const onloadModuleNameSuffix = "-onload-module"
+
+func kmmLabelName(name, namespace string) string {
+	return kmmModuleLabelPrefix + "." + namespace + "." + name + onloadModuleNameSuffix
+}
+
+const onloadLabelPrefix = "onload.amd.com/"
+
+func onloadLabelName(name, namespace string) string {
+	return onloadLabelPrefix + namespace + "." + name
+}
+
+func (r *OnloadReconciler) listNodesWithLabels(ctx context.Context, labelString string) (corev1.NodeList, error) {
+	selector, err := labels.Parse(labelString)
+	if err != nil {
+		return corev1.NodeList{}, err
+	}
+
+	nodes := corev1.NodeList{}
+	opt := client.MatchingLabelsSelector{Selector: selector}
+	err = r.List(ctx, &nodes, opt)
+	if err != nil {
+		return corev1.NodeList{}, err
+	}
+	return nodes, nil
+}
+
+func (r *OnloadReconciler) deleteLabels(ctx context.Context, namespacedName types.NamespacedName) error {
+	log := log.FromContext(ctx)
+
+	labelKey := kmmLabelName(namespacedName.Name, namespacedName.Namespace)
+
+	nodes, err := r.listNodesWithLabels(ctx, labelKey)
+	if err != nil {
+		return err
+	}
+
+	for _, node := range nodes.Items {
+		nodeCopy := node.DeepCopy()
+		delete(node.Labels, labelKey)
+		err := r.Patch(ctx, &node, client.MergeFrom(nodeCopy))
+		if err != nil {
+			log.Error(err, "Failed to patch node"+node.Name)
+			return err
+		}
+
+	}
+
+	labelKey = onloadLabelName(namespacedName.Name, namespacedName.Namespace)
+
+	nodes, err = r.listNodesWithLabels(ctx, labelKey)
+	if err != nil {
+		return err
+	}
+
+	for _, node := range nodes.Items {
+		nodeCopy := node.DeepCopy()
+		delete(node.Labels, labelKey)
+		err := r.Patch(ctx, &node, client.MergeFrom(nodeCopy))
+		if err != nil {
+			log.Error(err, "Failed to patch node"+node.Name)
+			return err
+		}
+
+	}
+	return nil
+}
+
+func (r *OnloadReconciler) addKmmLabelsToNodes(ctx context.Context, onload *onloadv1alpha1.Onload) (*ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	kmmPodLabelSet := labels.Set{"kmm.node.kubernetes.io/module.name": onload.Name + onloadModuleNameSuffix}
+
+	labels := labels.FormatLabels(onload.Spec.Selector)
+	labelKey := kmmLabelName(onload.Name, onload.Namespace)
+
+	nodes, err := r.listNodesWithLabels(ctx, labels)
+	if err != nil {
+		return nil, err
+	}
+
+	changesMade := false
+
+	for _, node := range nodes.Items {
+		if _, found := node.Labels[labelKey]; !found {
+			// Check that there isn't a lingering module pod
+			// This can be removed, but without it the upgrade process is more
+			// concurrent (rather than rolling), which leads to more pod
+			// restarts/failures. Both methods work, but this approach maintains
+			// the rolling nature of the update (as much as is possible).
+			pods, err := r.getPodsOnNode(ctx, kmmPodLabelSet, node.Name)
+			if err != nil {
+				return nil, err
+			} else if len(pods) > 0 {
+				log.Info("lingering module pod on node " + node.Name)
```

Comment:
```suggestion
				log.Info("Lingering module pod on node " + node.Name)
```

Also, whenever we return from here with the requeue request, we cause the "Labelled nodes with kmm label" log message, which isn't true. I haven't thought about how to address it other than returning `changesMade` .